src/main/java/com/Elessar/app/server/LogOnHandler.java

```diff
@@ -52,26 +56,39 @@ public void handle(HttpExchange he) throws IOException {
             final LogonRequest logonRequest = LogonRequest.parseFrom(is);
             final LogonResponse.Builder logonResponse = LogonResponse.newBuilder();
             final String userName = logonRequest.getName();
+            final String password = logonRequest.getPassword();
             final String clientURL = new StringBuilder().append("http://")
-                                                               .append(he.getRemoteAddress().getAddress().getHostAddress())
-                                                               .append(":")
-                                                               .append(logonRequest.getPort()).toString();
+                                                        .append(he.getRemoteAddress().getAddress().getHostAddress())
+                                                        .append(":")
+                                                        .append(logonRequest.getPort()).toString();
+
+            final User existingUser = users.getUnchecked(userName);
+            final User newUser = new User(userName, password, null, null, clientURL, true);
 
-            final User prevUser = db.update(new User(userName, logonRequest.getPassword(), null, null, clientURL, true));
+            if (existingUser.getName() == null || !password.equals(existingUser.getPassword())) {
+                // remove dummy entry as this is an invalid request
+                users.invalidate(userName);
 
-            if (prevUser == null) {
                 logger.info("User {} and password combination does NOT exist !", userName);
                 logonResponse.setSuccess(false).setFailReason("User " + userName + " password combination does NOT exist !");
                 he.sendResponseHeaders(400, 0);
-            } else if (prevUser.getOnline()) {
+            } else if (existingUser.getOnline()) {
                 logger.info("User {} has already log on !", userName);
                 logonResponse.setSuccess(true);
                 he.sendResponseHeaders(200, 0);
+
+                if (!clientURL.equals(existingUser.getURL())) {
+                    db.update(newUser);
+                    users.invalidate(userName);
```

Comment:
Just a thought but I can be wrong:
I guess after a user logs on, we would expect him to send / receives messages. So here instead of invalidating we could do a "users.put(userName, user)" to warm up the cache (or getUnchecked()) if possible.
Same in the else branch.

If it requires another query then we should stick to invalidate